### PR TITLE
support parallel nested nnet for Kaldi

### PR DESCRIPTION
**Pull request by pavel-esir**
Description: Changed **Slice** -> **AttributedVariadicSplit** while parsing ParalllelComponent (comes alongside with NestedNnet) in Kaldi loader.

JIRA: CVS-34431

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR: N/A -- no new transformations were added
* [x]  Transformation preserves original framework node names: N/A -- because no transformation were added

Validation:
* [x]  Unit tests: will be added in a separate PR https://github.com/openvinotoolkit/openvino/pull/1241
* [x]  Framework operation tests: N/A -- no new operations were added
* [x]  Transformation tests: N/A -- no new transformations were added
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py: N/A
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list: N/A
* [x]  Supported **public** models list: N/A
* [x]  New operations specification: N/A
* [x]  Guide on how to convert the **public** model: N/A
* [x]  User guide update: N/A

### CODE REVIEW

File: model-optimizer/mo/front/kaldi/extractors/slice_ext.py [hunk 1]
```diff
@@ -1,42 +0,0 @@
-"""
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/kaldi/extractors/slice_ext.py:1]: Why did you remove this extractor? Are you sure that this operation is not used in other models?

File: model-optimizer/mo/front/kaldi/loader/loader.py [hunk 2]
```diff
@@ -44,52 +43,52 @@ def load_parallel_component(file_descr, graph: Graph, prev_layer_id):
...
-        input_nodes = [n for n in graph.nodes(data=True) if n[1]['op'] == 'Parameter']
-        shape = input_nodes[0][1]['shape']
-        if i != nnet_count - 1:
-            slices_points.append(shape[1])
-        g.remove_node(input_nodes[0][0])
+
+        input_nodes = [n for n in g.nodes(data=True) if n[1]['op'] == 'Parameter']
+        if len(input_nodes) == 1:
+            shape = input_nodes[0][1]['shape']
+            split_points.append(shape[1])
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/kaldi/loader/loader.py:59]: Could you provide some comment in the code why do we use this value as a split points?

File: model-optimizer/mo/front/kaldi/loader/loader.py [hunk 3]
```diff
@@ -44,52 +43,52 @@ def load_parallel_component(file_descr, graph: Graph, prev_layer_id):
...
+            g.remove_node(input_nodes[0][0])
+        else:
+            raise Error('NestedNnet with multiple inputs is not supported')
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/kaldi/loader/loader.py:62]: Why don't we support multiple inputs? Is it a plugin/MO limitation?

File: model-optimizer/mo/front/kaldi/loader/loader.py [hunk 4]
```diff
@@ -44,52 +43,52 @@ def load_parallel_component(file_descr, graph: Graph, prev_layer_id):
...
+
         outputs.append(sorted_nodes[-1])
-    packed_sp = struct.pack("B", 4) + struct.pack("I", len(slices_points))
-    for i in slices_points:
-        packed_sp += struct.pack("I", i)
-    slice_node.parameters = io.BytesIO(packed_sp)
+        inputs.append(sorted_nodes[0])
+
+    split_id = graph.unique_id(prefix='NestedNets/Split')
+    mapping_rule = {'out_ports_count': nnet_count, 'size_splits': split_points, 'axis': 1, 'name': split_id}
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/kaldi/loader/loader.py:76]: nit: please, rename the variable `mapping_rule` to `attrs` or similar

File: model-optimizer/mo/front/kaldi/loader/loader.py [hunk 5]
```diff
@@ -44,52 +43,52 @@ def load_parallel_component(file_descr, graph: Graph, prev_layer_id):
...
+    split_node = AttributedVariadicSplit(graph, mapping_rule).create_node()
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/kaldi/loader/loader.py:77]: nit: variadic_split_node

File: model-optimizer/mo/front/kaldi/loader/loader.py [hunk 6]
```diff
@@ -44,52 +43,52 @@ def load_parallel_component(file_descr, graph: Graph, prev_layer_id):
...
+    Node(graph, prev_layer_id).add_output_port(0)
+    graph.create_edge(Node(graph, prev_layer_id), split_node, 0, 0)
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/kaldi/loader/loader.py:79]: Can we avoid using the deprecated API and instead of `create_edge` use `node.out_port(0).connect(...)`?

File: model-optimizer/mo/front/kaldi/extractors/slice_ext.py (hunk 1 above)
Comment (pavel-esir) [reply model-optimizer/mo/front/kaldi/extractors/slice_ext.py:1]: Before removing it I double checked. It's not used in other places. The only case when Slice nodes were created was for <ParallelComponent> which is now replaced by Split. Kaldi models itself don't have operation Slice. The old Slice was copypasted from Caffe.

File: model-optimizer/mo/front/kaldi/loader/loader.py (hunk 2 above)
Comment (pavel-esir) [reply model-optimizer/mo/front/kaldi/loader/loader.py]: Added comment in the new version

File: model-optimizer/mo/front/kaldi/loader/loader.py (hunk 3 above)
Comment (pavel-esir) [reply model-optimizer/mo/front/kaldi/loader/loader.py]: I checked sources of Kaldi, \<ParallelComponent\> which is present only in nnet1 supports only 1D inputs. This is limitations of Kaldi models not MO/plugin. This conditions newer happens, It was bad I removed it.

File: model-optimizer/mo/front/kaldi/loader/loader.py (hunk 5 above)
Comment (pavel-esir) [reply model-optimizer/mo/front/kaldi/loader/loader.py]: done

File: model-optimizer/mo/front/kaldi/loader/loader.py (hunk 4 above)
Comment (pavel-esir) [reply model-optimizer/mo/front/kaldi/loader/loader.py]: done

File: model-optimizer/mo/front/kaldi/loader/loader.py (hunk 6 above)
Comment (pavel-esir) [reply model-optimizer/mo/front/kaldi/loader/loader.py]: This is really a good questions. Firstly I tried use the newer API but faced the problem: for some reason it created additional data nodes. So I returned to old way. Especially since surrounding code was written in the old way.

After your comment I figured out why new API causes problems. When graph.stage == None (which is loader case) it creates data Nodes. I refreshed a couple of API functions so that they will not create data nodes by default. This should not break anything but let's wait for the tests.

Locally I am still getting correct graph which is inferred with high avg accuracy (1.03705e-09).

Comment (lazarevevgeny) [reply model-optimizer/mo/front/kaldi/loader/loader.py]: Ok.

**Comment by pavel-esir** [comment]
> I think it will be usefull to add unit test for parallel component loading in loader_test file

ok, I'll do it

File: model-optimizer/mo/front/kaldi/loader/loader.py (hunk 6 above)
Comment (pavel-esir) [reply model-optimizer/mo/front/kaldi/loader/loader.py]: If we change in `model-optimizer/mo/graph/graph.py:553` default stage to `'front`' half of unittests (936 of 1703) will fail. If we change behaviors of `add_destination` and `set_source` to not create by default data nodes. 
E.g. in `model-optimizer/mo/graph/connection.py:84`  change
`if self.graph.stage in ['front', None]:`
\# create without data nodes...
this will fix converting of the networks but ~100 unittests will fail.

Comment (lazarevevgeny) [reply model-optimizer/mo/front/kaldi/loader/loader.py]: Ok. Then leave it as is for now.

**Comment by pavel-esir** [comment]
Unittests will be added in a separate PR https://github.com/openvinotoolkit/openvino/pull/1241
CI checks have been completed,@lazarevevgeny, please merge.
